### DISCUSSION TITLE
Graceful shutdown disconnect clients & reconnection

### DISCUSSION THREAD

**Discussion by bartversluijs** [socketio/socket.io#4083] [Q&A]
Hi all,

I'm trying to implement a graceful shutdown in my project. I have the following setup:
```ts
for (const socket of this.socketio.sockets.sockets.values()) {
  // Disconnect (and destroy connection) for every connected socket
  socket.disconnect(true);
}

await new Promise<void>((resolve) => {
  this.socketio.close(() => resolve());
});
```

This works fine and it disconnects every client and gracefully shuts down the socket. Don't know if it's ideal, but it works. Only executing the `close()` function will close every connection but they reconnect again and the function will never perform a callback.

Only problem with this is that the clients will never try to reconnect after that, even with the `forceNew: true` option.
With multiple instances of the software, the clients should try to reconnect again so it'll reconnect with another instance. Now, it just stops connecting after one server went down.

I can probably handle this case in the client, but I believe this should be done by the `socket.io-client` package automatically.

Any suggestions?

Thanks in advance!

**Comment by bartversluijs** [comment]
It's been quite some time, but anyone has any idea how to solve this issue?
Another way to do it is force the Node.js process to close after some time and thus creating a Timeout functionality, but this is not ideal.

Thanks in advance!

**Comment by darrachequesne** [comment]
> Only executing the close() function will close every connection but they reconnect again and the function will never perform a callback

That sounds weird, the `close` method [here](https://github.com/socketio/socket.io/blob/9d86397243bcbb5775a29d96e5ef03e17148a8e7/lib/index.ts#L557) calls `httpServer.close()`, which should "stop the server from accepting new connections" according to the documentation [here](https://nodejs.org/api/http.html#serverclosecallback).

So the clients should not be able to reconnect directly. Could you please check?

**Reply by bartversluijs** [reply]
Thanks for your reply!
However, it's sharing the HTTP server with an Express instance, which is initiated to close after the socket.io is closed (different modules, so not really a way to close Express first). I'll try and debug it some more, but it seems like there's still a reconnection happening. Not always (unfortunately for the sake of debugging), but it happens...
